src/utils/whitespace-utils.js

```diff
@@ -0,0 +1,116 @@
+import escapeStringRegexp from "escape-string-regexp";
+
+class WhitespaceUtils {
+  #whitespaceCharacters;
+
+  constructor(whitespaceCharacters) {
+    this.#whitespaceCharacters = new Set(whitespaceCharacters);
+
+    if (
+      process.env.NODE_ENV !== "production" &&
+      (this.#whitespaceCharacters.size === 0 ||
+        Array.prototype.some.call(
+          whitespaceCharacters,
+          (character) => !/^\s$/.test(character)
+        ))
+    ) {
+      throw new TypeError(
+        `Invalid characters: ${JSON.stringify(whitespaceCharacters)}`
+      );
+    }
+  }
+
+  getLeadingWhitespaceCount(string) {
+    const whitespaceCharacters = this.#whitespaceCharacters;
+    let count = 0;
+
+    for (
+      let index = 0;
+      index < string.length && whitespaceCharacters.has(string.charAt(index));
+      index++
+    ) {
+      count++;
+    }
+
+    return count;
+  }
+
+  getTrailingWhitespaceCount(string) {
+    const whitespaceCharacters = this.#whitespaceCharacters;
+    let count = 0;
+
+    for (
+      let index = string.length - 1;
+      index >= 0 && whitespaceCharacters.has(string.charAt(index));
+      index--
+    ) {
+      count++;
+    }
+
+    return count;
+  }
+
+  getLeadingWhitespace(string) {
+    const count = this.getLeadingWhitespaceCount(string);
+    return string.slice(0, count);
+  }
+
+  getTrailingWhitespace(string) {
+    const count = this.getTrailingWhitespaceCount(string);
+    return string.slice(string.length - count);
+  }
+
+  hasLeadingWhitespace(string) {
+    return this.#whitespaceCharacters.has(string.charAt(0));
+  }
+
+  hasTrailingWhitespace(string) {
+    return this.#whitespaceCharacters.has(string.at(-1));
+  }
+
+  trimStart(string) {
+    const count = this.getLeadingWhitespaceCount(string);
+    return string.slice(count);
+  }
+
+  trimEnd(string) {
+    const count = this.getTrailingWhitespaceCount(string);
+    return string.slice(0, string.length - count);
+  }
+
+  trim(string) {
+    return this.trimEnd(this.trimStart(string));
+  }
+
+  split(string, captureWhitespace = false) {
```

Comment:
Should we name it as something like `splitByContinuousWhitespace` or `splitByLongestWhitespace`?